### PR TITLE
Remember the last used server for room directory searches

### PR DESCRIPTION
**Pull request by SimonBrandner**
Fixes https://github.com/vector-im/element-web/issues/17739

### CODE REVIEW

File: src/components/views/directory/NetworkDropdown.tsx [hunk 1]
```diff
@@ -41,7 +41,8 @@ import QuestionDialog from "../dialogs/QuestionDialog";
 import UIStore from "../../../stores/UIStore";
 import { compare } from "../../../utils/strings";
 
-export const ALL_ROOMS = Symbol("ALL_ROOMS");
+// XXX: We would ideally use a symbol here but we can't since we save this value to localStorage
```
Comment (t3chguy) [inline src/components/views/directory/NetworkDropdown.tsx:44]: You can, just call `String(symbol)` on it

Comment (t3chguy) [reply src/components/views/directory/NetworkDropdown.tsx:44]: oh but then rehydrating that symbol is a pain, nvm ignore this

Comment (SimonBrandner) [reply src/components/views/directory/NetworkDropdown.tsx:44]: But I would probably need some silly logic for reading from LS, right? `localStorage.getItem(key)` returns a string, so I would need to convert it to a symbol then but I would need to do this only if the the return value was `"ALL_ROOMS"`

File: src/components/structures/RoomDirectory.tsx [hunk 2]
```diff
@@ -150,8 +153,8 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
             publicRooms: [],
             loading: true,
             error: null,
-            instanceId: undefined,
-            roomServer: MatrixClientPeg.getHomeserverName(),
+            instanceId: localStorage.getItem(LAST_INSTANCE_KEY),
+            roomServer: localStorage.getItem(LAST_SERVER_KEY) || MatrixClientPeg.getHomeserverName(),
```
Comment (t3chguy) [inline src/components/structures/RoomDirectory.tsx:157]: what happens if that roomServer is no longer defined? e.g it was in config.json but is no longer there
what happens if that instanceId is no longer a valid one offered by the HS?

File: src/components/structures/RoomDirectory.tsx [hunk 3]
```diff
@@ -116,7 +119,30 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
...
+                    : myHomeserver;
+                const instanceIds = [];
+                if (roomServer === myHomeserver) {
+                    Object.values(this.protocols).forEach((protocol) => {
+                        protocol.instances.forEach((instance) => instanceIds.push(instance.instance_id));
+                    });
+                }
+                const instanceId = (instanceIds.includes(lsInstanceId) || lsInstanceId === ALL_ROOMS)
+                    ? lsInstanceId
+                    : null;
```
Comment (t3chguy) [inline src/components/structures/RoomDirectory.tsx:138]:
```suggestion
                let instanceId: string = null;
                if (lsInstanceId === ALL_ROOMS || (roomServer === myHomeserver && Object.values(this.protocols).some(p => p.instances.some(i => i.instance_id === lsInstanceId))) {
                	instanceId = lsInstanceId;
                }
```

with some line-splitting would be more readable IMO

File: src/components/structures/RoomDirectory.tsx [hunk 4]
```diff
@@ -116,7 +119,30 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
...
+
+                this.setState({
+                    protocolsLoading: false,
+                    instanceId: instanceId,
+                    roomServer: roomServer,
+                });
+                this.refreshRoomList();
```
Comment (t3chguy) [inline src/components/structures/RoomDirectory.tsx:145]: this is a bit annoying, could we not assume that the LS values are correct, load them in c'tor and then just validate them here, triggering a refresh only if the validation failed and they had to be reset

File: src/components/structures/RoomDirectory.tsx [hunk 5]
```diff
@@ -116,7 +119,30 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
...
+                    });
+                }
+                const instanceId = (instanceIds.includes(lsInstanceId) || lsInstanceId === ALL_ROOMS)
+                    ? lsInstanceId
+                    : null;
+
+                this.setState({
+                    protocolsLoading: false,
+                    instanceId: instanceId,
+                    roomServer: roomServer,
```
Comment (t3chguy) [inline src/components/structures/RoomDirectory.tsx:143]:
```suggestion
                    instanceId,
                    roomServer,
```

File: src/components/views/directory/NetworkDropdown.tsx [hunk 6]
```diff
@@ -171,7 +171,7 @@ const NetworkDropdown = ({ onOptionChange, protocols = {}, selectedServerName, s
 
             const protocolsList = server === hsName ? Object.values(protocols) : [];
             if (protocolsList.length > 0) {
-                // add a fake protocol with the ALL_ROOMS symbol
+                // add a fake protocol with the ALL_ROOMS
```
Comment (t3chguy) [inline src/components/views/directory/NetworkDropdown.tsx:174]:
```suggestion
                // add a fake protocol with ALL_ROOMS
```

File: src/components/structures/RoomDirectory.tsx [hunk 7]
```diff
@@ -116,7 +119,30 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
...
                 this.protocols = response;
-                this.setState({ protocolsLoading: false });
+                const myHomeserver = MatrixClientPeg.getHomeserverName();
+                const lsRoomServer = localStorage.getItem(LAST_SERVER_KEY);
+                const lsInstanceId = localStorage.getItem(LAST_INSTANCE_KEY);
+                const configSevers = SdkConfig.get().roomDirectory?.servers || [];
+                const settingsServers = SettingsStore.getValue("room_directory_servers") || [];
+                const roomServer = [...configSevers, ...settingsServers].includes(lsRoomServer)
+                    ? lsRoomServer
+                    : myHomeserver;
```
Comment (t3chguy) [inline src/components/structures/RoomDirectory.tsx:129]:
```suggestion

                const roomServer = (
                	SdkConfig.get().roomDirectory?.servers?.includes(lsRoomServer) ||
                	SettingsStore.getValue("room_directory_servers")?.includes(lsRoomServer)
                ) ? lsRoomServer : myHomeserver;
```

or similar is more readable IMO

File: src/components/structures/RoomDirectory.tsx [hunk 8]
```diff
@@ -122,27 +122,33 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
...
+                ) ? lsRoomServer : myHomeserver;
+
+                let instanceId: string = null;
+                if (
+                    (lsInstanceId === ALL_ROOMS) ||
+                    (
+                        roomServer === myHomeserver &&
+                        Object.values(this.protocols).some(p => p.instances.some(i => i.instance_id === lsInstanceId))
+                    )
+                ) {
```
Comment (t3chguy) [inline src/components/structures/RoomDirectory.tsx:137]: actually, I misled you!

`roomServer === myHomeserver` is needed for both cases

```suggestion
                if (roomServer === myHomeserver && (
                	lsInstanceId === ALL_ROOMS) ||
                    Object.values(this.protocols).some(p => p.instances.some(i => i.instance_id === lsInstanceId))
                )) {
```

File: src/components/structures/RoomDirectory.tsx [hunk 9]
```diff
@@ -122,27 +122,33 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
...
-                    : myHomeserver;
-                const instanceIds = [];
-                if (roomServer === myHomeserver) {
-                    Object.values(this.protocols).forEach((protocol) => {
-                        protocol.instances.forEach((instance) => instanceIds.push(instance.instance_id));
-                    });
+                const roomServer = (
+                    SdkConfig.get().roomDirectory?.servers?.includes(lsRoomServer) ||
+                    SettingsStore.getValue("room_directory_servers")?.includes(lsRoomServer)
+                ) ? lsRoomServer : myHomeserver;
```
Comment (t3chguy) [inline src/components/structures/RoomDirectory.tsx:128]: Hmm maybe even

```suggestion
				let roomServer = myHomeserver;
				if (
                    SdkConfig.get().roomDirectory?.servers?.includes(lsRoomServer) ||
                    SettingsStore.getValue("room_directory_servers")?.includes(lsRoomServer)
                ) {
                	roomServer = lsRoomServer;
                }
```

File: src/components/structures/RoomDirectory.tsx (hunk 8 above)
Comment (SimonBrandner) [reply src/components/structures/RoomDirectory.tsx:137]: /me should actually read the code before copy-pasting it...

**Review by t3chguy — APPROVED** [review]
LGTM, thanks!! :shipit:

**Pull request merged**